pkg/icingadb/history/sync.go

```diff
@@ -32,274 +35,366 @@ func NewSync(db *icingadb.DB, redis *icingaredis.Client, logger *zap.SugaredLogg
-	structifiers []structify.MapStructifier
-}
+// userNotificationStage is a specialized stageFunc that populates the user_notification_history table. It is executed
+// on the notification history stream and uses the users_notified_ids attribute to create an entry in the
+// user_notification_history relation table for each user ID.
+func userNotificationStage(ctx context.Context, s Sync, key string, in <-chan redis.XMessage, out chan<- redis.XMessage) error {
+	type NotificationHistory struct {
+		Id            types.UUID   `structify:"id"`
+		EnvironmentId types.Binary `structify:"environment_id"`
+		EndpointId    types.Binary `structify:"endpoint_id"`
+		UserIds       string       `structify:"users_notified_ids"`
+	}
+
+	structifier := structify.MakeMapStructifier(reflect.TypeOf((*NotificationHistory)(nil)).Elem(), "structify")
 
-// historyStreams contains all Redis history streams to sync.
-var historyStreams = func() []historyStream {
-	var streams []historyStream
-	for _, rhs := range []struct {
-		kind       string
-		structPtrs []v1.UpserterEntity
-	}{
-		{"notification", []v1.UpserterEntity{(*v1.NotificationHistory)(nil), (*v1.HistoryNotification)(nil)}},
-		{"usernotification", []v1.UpserterEntity{(*v1.UserNotificationHistory)(nil)}},
-		{"state", []v1.UpserterEntity{(*v1.StateHistory)(nil), (*v1.HistoryState)(nil)}},
-		{"downtime", []v1.UpserterEntity{(*v1.DowntimeHistory)(nil), (*v1.HistoryDowntime)(nil)}},
-		{"comment", []v1.UpserterEntity{(*v1.CommentHistory)(nil), (*v1.HistoryComment)(nil)}},
-		{"flapping", []v1.UpserterEntity{(*v1.FlappingHistory)(nil), (*v1.HistoryFlapping)(nil)}},
-		{"acknowledgement", []v1.UpserterEntity{(*v1.AcknowledgementHistory)(nil), (*v1.HistoryAck)(nil)}},
-	} {
-		var structifiers []structify.MapStructifier
-		for _, structPtr := range rhs.structPtrs {
-			structifiers = append(structifiers, structify.MakeMapStructifier(reflect.TypeOf(structPtr).Elem(), "json"))
+	baseUUID := uuid.MustParse("235396c3-a525-4806-a3bc-3806c25baaeb")
+
+	return writeMultiEntityStage(func(entry redis.XMessage) ([]v1.UpserterEntity, error) {
+		rawNotificationHistory, err := structifier(entry.Values)
+		if err != nil {
+			return nil, err
 		}
+		notificationHistory := rawNotificationHistory.(*NotificationHistory)
 
-		streams = append(streams, historyStream{rhs.kind, structifiers})
-	}
+		var users []types.Binary
+		err = json.Unmarshal([]byte(notificationHistory.UserIds), &users)
+		if err != nil {
+			return nil, err
+		}
 
-	return streams
-}()
+		var userNotifications []v1.UpserterEntity
+
+		for _, user := range users {
+			userNotifications = append(userNotifications, &v1.UserNotificationHistory{
+				HistoryTableEntity: v1.HistoryTableEntity{
+					// Generate a deterministic ID based on the other fields.
+					Id: types.UUID{UUID: uuid.NewSHA1(baseUUID, append(append(append([]byte(nil),
+						notificationHistory.EnvironmentId...), notificationHistory.Id.UUID[:]...), user...))},
```

Comment:
notificationHistory.EnvironmentId? Isn't notificationHistory.Id unique enough?

Reply:
user? The whole user ID? Wouldn't i be enough? (i from `s/for _, user/for i, user/`)

Reply:
What would be the benefit? Apart from that you'd have to sort the user list to generate consistent IDs on both instances in a HA setup (if that's what we end up doing). Apart from that, most from the explanation in https://github.com/Icinga/icingadb/pull/364#discussion_r716819291 also applies here.

Reply:
Less stuff to hash. Good for performance and checksums quality.

Reply:
SHA1 operates on 512 bit blocks, so it doesn't matter if I put in 448 bits (current code), 352 bits (use 64 bits for uid) or 320 bits (32 bits for uid).